### PR TITLE
Added messaging for multiple users and new feature user's group

### PR DESCRIPTION
**Pull request by captjt**
* `go fmt` on DiscardStats methods

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1 +1,2 @@
 *.sw[o-p]
+vendor
```
Comment (briankassouf) [inline .gitignore]: We don't want to ignore the vendor directory. Eventually (in another PR) all the vendored packages should be updated. I think this will require some code updates too, but shouldn't be too bad.

File: Dockerfile [hunk 2]
```diff
@@ -1,9 +1,8 @@
-FROM golang:1.5
-
+FROM golang:1.8
 RUN mkdir /etc/incus
 
-ADD . /go/src/github.com/Imgur/incus
-WORKDIR /go/src/github.com/Imgur/incus
+ADD . /go/src/github.com/jtaylor32/incus
+WORKDIR /go/src/github.com/jtaylor32/incus
```
Comment (briankassouf) [inline Dockerfile]: Can you be sure to change this back? When working on a fork with Golang I find this handy: https://stackoverflow.com/questions/14323872/using-forked-package-import-in-go

File: README.md [hunk 3]
```diff
@@ -213,7 +213,7 @@ sudo /etc/init.d/incus restart
 ## Configuration
 Incus needs to be restarted after any configuration change.
 
-#### CLIENT_BROADCASTS
+#### ClientBroadcast
```
Comment (briankassouf) [inline README.md]: This should be changed back as the configuration option should still be the underscore'd version. see config.go for more info.

File: glide.lock [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+  subpackages:
+  - unix
+- name: golang.org/x/text
+  version: f28f36722d5ef2f9655ad3de1f248e3e52ad5ebd
+  subpackages:
+  - transform
+  - unicode/norm
+- name: gopkg.in/yaml.v2
+  version: a3f3340b5840cee44f372bddb5880fcbc419b46a
+testImports: []
```
Comment (briankassouf) [inline glide.lock]: The current vendor tool is Godeps. Any major vendor changes should be done in another PR. (Anything that doesn't pertain to the groups feature)

File: incus/main.go [hunk 5]
```diff
@@ -11,7 +11,7 @@ import (
 	"syscall"
 	"time"
 
-	"github.com/Imgur/incus"
+	"github.com/jtaylor32/incus"
```
Comment (briankassouf) [inline incus/main.go]: Same, please make sure this is changed back

File: scripts/install.sh [hunk 6]
```diff
@@ -1,6 +1,6 @@
 #!/bin/bash
 
-GOPROJ=$GOPATH/src/github.com/Imgur/incus
+GOPROJ=$GOPATH/src/github.com/jtaylor32/incus
```
Comment (briankassouf) [inline scripts/install.sh]: Same as above.

File: store.go [hunk 7]
```diff
@@ -132,3 +132,37 @@ func (this *Storage) getPage(page string) map[string]*Socket {
 	this.pageMu.RLock()
 	return this.memory.getPage(page)
 }
+
+func (this *Storage) SetGroups(sock *Socket) error {
+	this.pageMu.Lock()
```
Comment (briankassouf) [inline store.go]: Can you add a new lock to the storage struct? It should use a separate lock for the groups map.

File: store.go [hunk 8]
```diff
@@ -132,3 +132,37 @@ func (this *Storage) getPage(page string) map[string]*Socket {
...
+		if err := this.redis.SetGroups(sock); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (this *Storage) UnsetGroups(sock *Socket) error {
+	this.pageMu.Lock()
```
Comment (briankassouf) [inline store.go]: Same, it should use a different lock.

File: store.go [hunk 9]
```diff
@@ -132,3 +132,37 @@ func (this *Storage) getPage(page string) map[string]*Socket {
...
+	this.memory.UnsetGroups(sock)
+	this.pageMu.Unlock()
+
+	if this.StorageType == "redis" {
+		if err := this.redis.UnsetPage(sock); err != nil {
```
Comment (briankassouf) [inline store.go]: `UnsetGroup`

File: store.go [hunk 10]
```diff
@@ -132,3 +132,37 @@ func (this *Storage) getPage(page string) map[string]*Socket {
...
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (this *Storage) getGroups(groups []string) []map[string]*Socket {
+	defer this.pageMu.RUnlock()
```
Comment (briankassouf) [inline store.go]: Same, should use a different lock. Also could you move the defer statement below the `Lock()` call?

File: message.go [hunk 11]
```diff
@@ -338,6 +375,74 @@ func (this *CommandMsg) messageUser(UID string, page string, server *Server) {
...
+			}
+
+			select {
+			default:
+			case <-ctx.Done():
+				return ctx.Err()
+			}
+		}
+		return nil
+	})
```
Comment (briankassouf) [inline message.go:440]: Rather than spawn two more go routines, what if it looped through the users and called `this.messageUser(uid, page, server)`?

File: message.go [hunk 12]
```diff
@@ -380,7 +485,31 @@ func (this *CommandMsg) messagePage(page string, server *Server) {
...
 
+func (this *CommandMsg) messageGroups(groups []string, server *Server) {
+	msg, err := this.formatMessage()
+	if err != nil {
+		return
+	}
+
+	server.Stats.LogPageMessage()
+
+	groupMap := server.Store.getGroups(groups)
```
Comment (briankassouf) [inline message.go]: This is technically a `groupList`

File: message.go [hunk 13]
```diff
@@ -61,6 +63,22 @@ func (this *CommandMsg) FromSocket(sock *Socket) {
...
+		grps, ok := this.Command["groups"]
+		grps = strings.TrimSpace(strings.Replace(grps, " ", "", -1))
+		groups := strings.Split(grps, ",")
+
+		if !ok || len(groups) == 0 {
+			return
+		}
+
+		// hack check for empty slice
+		if len(sock.Groups) != 0 {
```
Comment (briankassouf) [inline message.go]: Should this be moved above the `len(groups) == 0` so that an empty group command could remove all the socket from all groups?

File: stats.go [hunk 14]
```diff
@@ -39,7 +40,7 @@ type DiscardStats struct{}
 
 func (d *DiscardStats) LogStartup()                                   {}
 func (d *DiscardStats) LogClientCount(int64)                          {}
-func (d *DiscardStats) LogGoroutines(int)                          {}
+func (d *DiscardStats) LogGoroutines(int)                             {}
 func (d *DiscardStats) LogCommand(from, cmdType string)               {}
 func (d *DiscardStats) LogPageMessage()                               {}
 func (d *DiscardStats) LogUserMessage()                               {}
```
Comment (briankassouf) [inline stats.go]: Could you add a `LogGroupMessage()` stat function?

File: message.go [hunk 15]
```diff
@@ -380,7 +485,31 @@ func (this *CommandMsg) messagePage(page string, server *Server) {
 	return
 }
 
+func (this *CommandMsg) messageGroups(groups []string, server *Server) {
+	msg, err := this.formatMessage()
+	if err != nil {
+		return
+	}
+
+	server.Stats.LogPageMessage()
```
Comment (briankassouf) [inline message.go]: This should be calling a `LogGroupMessage()` function and should probably take in the length of the group array. so something like: `LogGroupMessage(len(groups))`

**Review by briankassouf — CHANGES_REQUESTED** [review]
Overall really awesome! Just a few minor things and nitpicks. Thanks for adding this feature!

**Comment by briankassouf** [comment]
Also could you update the title of the PR to reflect the added features? And look into adding some test cases for the groups feature?

File: stats.go (hunk 14 above)
Comment (captjt) [reply stats.go]: Absolutely. I apologize at first it was only the initial formatting. Just kept adding stuff while I went along. I will clean everything and make sure it's more clear of what changes where made. 

Thanks!

File: Dockerfile (hunk 2 above)
Comment (captjt) [reply Dockerfile]: Perfect great source. Thanks!

File: message.go (hunk 13 above)
Comment (captjt) [reply message.go]: What I was trying to do is add groups to a single socket. So let's say you have your UID and you are apart of 3 different groups. I have mine and I'm in 2 groups. Someone sends you a message and to one of my groups. We both get that message kind of thing. Without too much knowledge of Redis I've been messing with just clearing them when setting new groups. Will gladly take recommendations on how you'd implement it.

File: message.go (hunk 11 above)
Comment (captjt) [reply message.go:440]: My thought process here was messageUsers will actually take a single User as well similar to messageUser (removing the messageUser function). The messaging process with the concurrency of two goroutines seems nicer at scale if someone was sending many-many messages with Incus. If you think that a simple loop would be faster I can change that as well.

File: message.go (hunk 15 above)
Comment (captjt) [reply message.go]: I will get the logging function finished and pushed up today.

File: glide.lock (hunk 4 above)
Comment (captjt) [reply glide.lock]: @briankassouf do you have any preference whether or not I open a PR on using a different tool? This raising the questions of how you'd like to handle the vendor directory as well. We can address this after I clean up this PR.

**Comment by jtaylor32** [comment]
Will rebase and squash to clean up git history. 🙈

**Comment by captjt** [comment]
I removed the "groups" concept for right now. User's can have a single _group_ similar to a _page_. This is because it will be a bigger feature for a different PR. If you have a group of friends you want to chat with while that's separate from page-to-page you can set a "group" for chatting that way.

**Comment by captjt** [comment]
@briankassouf with what I made some changes to what else do you think needs to be tweaked?
